Index: util.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- util.c	(revision 53bc57512c19b11eebc403a4cb2bbf7295eb0db1)
+++ util.c	(date 1584719903137)
@@ -18,7 +18,9 @@
 #include <strings.h>
 #include <stdlib.h>
 #include <stdio.h>
+#ifndef _MSC_VER
 #include <sys/time.h>
+#endif
 #include <time.h>
 #include <ctype.h>
 
Index: ldns/resolver.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- ldns/resolver.h	(revision 53bc57512c19b11eebc403a4cb2bbf7295eb0db1)
+++ ldns/resolver.h	(date 1584718953801)
@@ -20,13 +20,16 @@
 #ifndef LDNS_RESOLVER_H
 #define LDNS_RESOLVER_H
 
+#include <ldns/config.h>
 #include <ldns/error.h>
 #include <ldns/common.h>
 #include <ldns/rr.h>
 #include <ldns/tsig.h>
 #include <ldns/rdata.h>
 #include <ldns/packet.h>
+#ifndef _MSC_VER
 #include <sys/time.h>
+#endif
 
 #ifdef __cplusplus
 extern "C" {
Index: net.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- net.c	(revision 53bc57512c19b11eebc403a4cb2bbf7295eb0db1)
+++ net.c	(date 1581415335811)
@@ -27,7 +27,9 @@
 #ifdef HAVE_ARPA_INET_H
 #include <arpa/inet.h>
 #endif
+#ifndef _MSC_VER
 #include <sys/time.h>
+#endif
 #include <errno.h>
 #include <fcntl.h>
 #ifdef HAVE_POLL
Index: ldns/packet.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- ldns/packet.h	(revision 53bc57512c19b11eebc403a4cb2bbf7295eb0db1)
+++ ldns/packet.h	(date 1584718969611)
@@ -32,10 +32,13 @@
 #define LDNS_RA		32      /* Recursion Available - server flag */
 #define LDNS_AD		64      /* Authenticated Data - server flag */
 
+#include <ldns/config.h>
 #include <ldns/error.h>
 #include <ldns/common.h>
 #include <ldns/rr.h>
+#ifndef _MSC_VER
 #include <sys/time.h>
+#endif
 
 #ifdef __cplusplus
 extern "C" {
Index: host2str.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- host2str.c	(revision 53bc57512c19b11eebc403a4cb2bbf7295eb0db1)
+++ host2str.c	(date 1584720299830)
@@ -26,7 +26,9 @@
 #include <netdb.h>
 #endif
 #include <time.h>
+#ifndef _MSC_VER
 #include <sys/time.h>
+#endif
 
 #ifdef HAVE_SSL
 #include <openssl/bn.h>
