Index: evutil.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- evutil.c	(revision c7da1ef8e4057d87c39dde76dda8fabced4456d8)
+++ evutil.c	(date 1600188716344)
@@ -1738,7 +1738,9 @@
 	if (0 != FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM |
 			       FORMAT_MESSAGE_IGNORE_INSERTS |
 			       FORMAT_MESSAGE_ALLOCATE_BUFFER,
-			       NULL, errcode, 0, (char *)&msg, 0, NULL))
+			       NULL, errcode,
+			       MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US),
+			       (char *)&msg, 0, NULL))
 		chomp (msg);	/* because message has trailing newline */
 	else {
 		size_t len = 50;
