diff --git a/cmake/conan_provider.cmake b/cmake/conan_provider.cmake
index 1e50e69..c109e23 100644
--- a/cmake/conan_provider.cmake
+++ b/cmake/conan_provider.cmake
@@ -87,6 +87,10 @@ function(detect_arch ARCH)
         set(_ARCH x86)
     elseif(host_arch MATCHES "AMD64|amd64|x86_64|x64")
         set(_ARCH x86_64)
+    elseif(host_arch MATCHES "mipsel")
+        set(_ARCH mipsel)
+    elseif(host_arch MATCHES "mips")
+        set(_ARCH mips)
     endif()
     message(STATUS "CMake-Conan: cmake_system_processor=${_ARCH}")
     set(${ARCH} ${_ARCH} PARENT_SCOPE)
@@ -387,8 +391,8 @@ function(conan_install)
     set(CONAN_OUTPUT_FOLDER ${CMAKE_BINARY_DIR}/conan)
     # Invoke "conan install" with the provided arguments
     set(CONAN_ARGS ${CONAN_ARGS} -of=${CONAN_OUTPUT_FOLDER})
-    message(STATUS "CMake-Conan: conan install ${CMAKE_SOURCE_DIR} ${CONAN_ARGS} ${ARGN}")
-    execute_process(COMMAND ${CONAN_COMMAND} install ${CMAKE_SOURCE_DIR} ${CONAN_ARGS} ${ARGN} --format=json
+    message(STATUS "CMake-Conan: conan install ${CMAKE_CURRENT_FUNCTION_LIST_DIR}/.. ${CONAN_ARGS} ${ARGN}")
+    execute_process(COMMAND ${CONAN_COMMAND} install ${CMAKE_CURRENT_FUNCTION_LIST_DIR}/.. ${CONAN_ARGS} ${ARGN} --format=json
                     RESULT_VARIABLE return_code
                     OUTPUT_VARIABLE conan_stdout
                     ERROR_VARIABLE conan_stderr
@@ -407,8 +411,8 @@ function(conan_install)
         set_property(GLOBAL PROPERTY CONAN_GENERATORS_FOLDER "${CONAN_GENERATORS_FOLDER}")
         # reconfigure on conanfile changes
         string(JSON CONANFILE GET ${conan_stdout} graph nodes 0 label)
-        message(STATUS "CMake-Conan: CONANFILE=${CMAKE_SOURCE_DIR}/${CONANFILE}")
-        set_property(DIRECTORY ${CMAKE_SOURCE_DIR} APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${CMAKE_SOURCE_DIR}/${CONANFILE}")
+        message(STATUS "CMake-Conan: CONANFILE=${CMAKE_CURRENT_FUNCTION_LIST_DIR}/../${CONANFILE}")
+        set_property(DIRECTORY ${CMAKE_SOURCE_DIR} APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/../${CONANFILE}")
         # success
         set_property(GLOBAL PROPERTY CONAN_INSTALL_SUCCESS TRUE)
     endif()
